### PR TITLE
schemachanger: add end-to-end testing framework

### PR DESCRIPTION
**Pull request by postamar**
Recently we modified the declarative schema changer to be able to inject
all dependencies. This commit follows up on that work by adopting this
DI in new, data-driven, end-to-end tests of the declarative schema
changer.

Release note: None

### CODE REVIEW

File: pkg/sql/schemachanger/end_to_end_test.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	return execInCluster(ctx, t, tdb, stmt)
+}
+
+// waitForSchemaChangeJobsCompletion waits for all old schema changer jobs to
+// complete.
+func waitForSchemaChangeJobsCompletion(t *testing.T, tdb *sqlutils.SQLRunner) {
+	const q = `
+		SELECT count(*)
+		FROM [SHOW JOBS]
+		WHERE job_type IN ('SCHEMA CHANGE', 'TYPEDESC SCHEMA CHANGE', 'NEW SCHEMA CHANGE')
```
Comment (ajwerner) [inline pkg/sql/schemachanger/end_to_end_test.go:181]: nit: use the constants?

File: pkg/sql/schemachanger/end_to_end_test.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/datadriven"
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v2"
+)
+
+func TestEndToEnd(t *testing.T) {
```
Comment (ajwerner) [inline pkg/sql/schemachanger/end_to_end_test.go:45]: Commentary here feels worthwhile

Comment (postamar) [reply pkg/sql/schemachanger/end_to_end_test.go:45]: Note: this PR is still in draft mode, it's not done yet. I want to add the end-to-end-with-side-effects tests I told you about the other day.

File: pkg/sql/schemachanger/scexec/exec_mutation.go [hunk 3]
```diff
@@ -70,7 +67,7 @@ func executeDescriptorMutationOps(ctx context.Context, deps Dependencies, ops []
 
 type mutationVisitorState struct {
 	c                     Catalog
-	checkedOutDescriptors catalog.DescriptorIDSet
+	checkedOutDescriptors nstree.Map
```
Comment (postamar) [inline pkg/sql/schemachanger/scexec/exec_mutation.go:70]: I need this to be able to log side effects. Ultimately this is also what prompted the introduction of `descBuilder` and friends.

File: pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go [hunk 4]
```diff
@@ -489,42 +502,109 @@ func (b *testCatalogChangeBatcher) DeleteName(
...
+			panic(err)
+		}
+		str, err := json.Pretty(js)
+		if err != nil {
+			panic(err)
+		}
+		return str
+	}
+
+	d := diff.Diff(toJson(a), toJson(b))
```
Comment (postamar) [inline pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go:555]: This is where we use the vendored dependency I'm adding. I couldn't find a better way to diff JSON objects than via their text representation. I think it's fine in practice.

File: pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go [hunk 5]
```diff
@@ -489,42 +502,109 @@ func (b *testCatalogChangeBatcher) DeleteName(
...
 func (b *testCatalogChangeBatcher) ValidateAndRun(ctx context.Context) error {
-	for nameInfo, id := range b.namesToDelete {
+	names := make([]descpb.NameInfo, 0, len(b.namesToDelete))
+	for nameInfo := range b.namesToDelete {
+		names = append(names, nameInfo)
+	}
+	sort.Slice(names, func(i, j int) bool {
+		return b.namesToDelete[names[i]] < b.namesToDelete[names[j]]
+	})
+	for _, nameInfo := range names {
```
Comment (postamar) [inline pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go:512]: We now need this order to be stable because it impacts the `side_effects` test.

**Review by postamar — COMMENTED** [review]
This should be ready for review, assuming that the CI results don't turn out to be too embarrassing.

**Review by ajwerner — COMMENTED** [review]
> Reviewed 14 of 17 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72283)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/schemachanger/end_to_end_test.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-MoE_2qa3ORLbR2bpU-J:-MoE_2qa3ORLbR2bpU-K:b-eo3vkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/85032404bbaf56b396d9fd2c83026c345129c7fe/pkg/sql/schemachanger/end_to_end_test.go#L73)):*
> ```Go
> 	} {
> 		t.Run(mode.name, func(t *testing.T) {
> 			runDataDrivenSchemaChangeTest(ctx, t, "catalog_state", mode.exec, marshalCatalog)
> ```

do you like having separate tests for catalog state and for side-effects? It feels weird to want to duplicate the tests statements. Catalog state in particular feels independent of the test itself. I think my preference would be to have `test` execute the statements and print the side-effects, ideally also printing the phase and stage in which they occurred and then separately have a `catalog-state` command which prints the catalog state. 

Honestly, for the catalog state, it feels like it'd be nice to have some diffing on that as well. Maybe we should add a primitive to label a catalog state and then be able to diff two of them. 
___
*[pkg/sql/schemachanger/end_to_end_test.go, line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-MoEYYEF021nxuYWucm3:-MoEYYEF021nxuYWucm4:b9biwqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/85032404bbaf56b396d9fd2c83026c345129c7fe/pkg/sql/schemachanger/end_to_end_test.go#L161)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 		})
> 
> 		// Run pre-commit phase.
> 		var jobID jobspb.JobID
> 		deps.WithTxn(scop.PreCommitPhase, func(s *sctestdeps.TestState) {
> ```
</details>

Is there a reason to not run these in the same transaction? In the "real world" they are in the same transaction. If we want to support more than one statement (which we will), then we should run all of them in the same transaction. I do get that maybe it doesn't matter because this doesn't do much, but it seems better to make this reflect reality, no?
___
*[pkg/sql/schemachanger/end_to_end_test.go, line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-MoEYnubA-aiobvI7jQ3:-MoEYnubA-aiobvI7jQ4:b9dlcav) ([raw file](https://github.com/cockroachdb/cockroach/blob/85032404bbaf56b396d9fd2c83026c345129c7fe/pkg/sql/schemachanger/end_to_end_test.go#L194)):*
> ```
> 	// Return a test state built _after_ running the DDL statement.
> 	return sctestdeps.NewTestDependencies(ctx, t, tdb, nil /* testingKnobs */, []string{stmt.SQL})
> ```
Can you say more about why we want it after? It's barely something I understand now, I certainly won't understand it in 6 months time without a lot of paging stuff back in.

**Review by postamar — COMMENTED** [review]
Thanks for taking a look! I'll follow up with something nicer soon.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-:-MoEanh83E20mVk3TkUk:b-yfcl80)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schemachanger/end_to_end_test.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-MnR3he5TDPBkUsu6bO_-r1-181:-MoEajcRB88iUnzQA5Xo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f59258ff178ced1fab2c7b976c2be222dfdff99d/pkg/sql/schemachanger/end_to_end_test.go#L181)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: use the constants?
</blockquote></details>

Done.
___
*[pkg/sql/schemachanger/end_to_end_test.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-MoE_2qa3ORLbR2bpU-J:-MoEa9MJEfIYC9pF-ZuM:b-3d70gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/85032404bbaf56b396d9fd2c83026c345129c7fe/pkg/sql/schemachanger/end_to_end_test.go#L73)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

do you like having separate tests for catalog state and for side-effects? It feels weird to want to duplicate the tests statements. Catalog state in particular feels independent of the test itself. I think my preference would be to have `test` execute the statements and print the side-effects, ideally also printing the phase and stage in which they occurred and then separately have a `catalog-state` command which prints the catalog state. 

Honestly, for the catalog state, it feels like it'd be nice to have some diffing on that as well. Maybe we should add a primitive to label a catalog state and then be able to diff two of them. 
</blockquote></details>

You're right, this won't age well at all. I like all of your suggestions here, I'll follow up on them.
___
*[pkg/sql/schemachanger/end_to_end_test.go, line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-MoEYYEF021nxuYWucm3:-MoEaKL59bZH9IqhOaLb:bo4vdwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/85032404bbaf56b396d9fd2c83026c345129c7fe/pkg/sql/schemachanger/end_to_end_test.go#L161)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		})
> 
> 		// Run pre-commit phase.
> 		var jobID jobspb.JobID
> 		deps.WithTxn(scop.PreCommitPhase, func(s *sctestdeps.TestState) {
> ```
Is there a reason to not run these in the same transaction? In the "real world" they are in the same transaction. If we want to support more than one statement (which we will), then we should run all of them in the same transaction. I do get that maybe it doesn't matter because this doesn't do much, but it seems better to make this reflect reality, no?
</blockquote></details>

No good reason! You're right.
___
*[pkg/sql/schemachanger/end_to_end_test.go, line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-MoEYnubA-aiobvI7jQ3:-MoEaWjDBMIF8jEIT7DL:btohwuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/85032404bbaf56b396d9fd2c83026c345129c7fe/pkg/sql/schemachanger/end_to_end_test.go#L194)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Return a test state built _after_ running the DDL statement.
> 	return sctestdeps.NewTestDependencies(ctx, t, tdb, nil /* testingKnobs */, []string{stmt.SQL})
> ```
Can you say more about why we want it after? It's barely something I understand now, I certainly won't understand it in 6 months time without a lot of paging stuff back in. 
</blockquote></details>

I'll add more commentary and improve this area. There's no good reason for it to be confusing.

**Comment by postamar** [comment]
I trimmed down the tests to just the side effects, because, as it turns out, serializing descriptors to json or yaml isn't stable. Well, the serialization is. obviously, but descriptors have arrays whose contents aren't in a stable order, `DependsOnTypes` being one of them for example.

**Comment by postamar** [comment]
I rebased and took advantage of @fqazi 's recent addition of CREATE INDEX to add another side-effects test suite. I also improved the side-effects testing framework by putting namespace changes in the expected results for `setup`, which makes the side-effects output easier to follow.

In the process of doing this I noticed a few sources of noise in `scmutationexec`, which I fixed in a subsequent commit. I also got annoyed at the long and varied variable names in that package so I pushed another commit as well.

I'm quite happy with how things are right now, please take another look.

**Review by fqazi — COMMENTED** [review]
This looks really good!

> Reviewed 3 of 9 files at r4, 13 of 13 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-:-MoKtXPKCdBwYe58kVql:b-9tsrt4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

**Review by fqazi — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72283#-:-MoKtb22Dj5kCHAWHcnl:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @fqazi)

**Review by ajwerner — APPROVED** [review]
LGTM

**Pull request merged**